File: Find_K_Pairs_with_Smallest_Sums.js
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @param {number} k
 * @return {number[][]}
 */
var kSmallestPairs = function(nums1, nums2, k) {
    const minHeap = new MinPriorityQueue({ priority: x => x[0] });
    
    for (let i = 0; i < nums1.length; i++) {
        const num1 = nums1[i];
        const num2 = nums2[0];
        
        minHeap.enqueue([num1 + num2, i, 0]);
    }
    
    const n = nums2.length;
    const res = [];
    
    while (k > 0 && !minHeap.isEmpty()) {
        const [sum, idx1, idx2] = minHeap.dequeue().element;
        
        res.push([nums1[idx1], nums2[idx2]]);
        
        if (res.length === k) return res;
        
        if (idx2 < n - 1) {
            minHeap.enqueue([nums1[idx1] + nums2[idx2 + 1], idx1, idx2 + 1]);
        } 
    }
    
    return res;   
};

File: Find_K_Pairs_with_Smallest_Sums.py
class Solution:
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        heap = []

        for i in range(len(nums1)):
            heapq.heappush(heap, (nums1[i] + nums2[0], nums1[i], nums2[0], 0))
        
        res = []
        while heap and len(res) < k:
            _, num1, num2, idx = heapq.heappop(heap)
            res.append([num1, num2])

            if idx + 1 < len(nums2):
                heapq.heappush(heap, (num1 + nums2[idx+1], num1, nums2[idx+1], idx+1))
        
        return res
